pkg/operator/ceph/cluster/mon/health.go

```diff
@@ -459,38 +533,52 @@ func (c *Cluster) updateMonDeploymentReplica(name string, enabled bool) error {
 func (c *Cluster) failoverMon(name string) error {
 	logger.Infof("Failing over monitor %q", name)
 
+	// remove the failed mon from a local list of the existing mons for finding a stretch zone
+	existingMons := c.clusterInfoToMonConfigWithExclude(name)
+
+	zone, err := c.findAvailableZoneIfStretched(existingMons)
+	if err != nil {
+		return errors.Wrap(err, "failed to find available stretch zone")
+	}
+
+	// Start a new monitor
+	m := c.newMonConfig(c.maxMonID+1, zone)
+	logger.Infof("starting new mon: %+v", m)
+
 	// Scale down the failed mon to allow a new one to start
 	if err := c.updateMonDeploymentReplica(name, false); err != nil {
 		// attempt to continue with the failover even if the bad mon could not be stopped
 		logger.Warningf("failed to stop mon %q for failover. %v", name, err)
 	}
+
+	// If the mon failover is not successful, revert the failover
 	newMonSucceeded := false
+	newMonMightBeInQuorum := false
 	defer func() {
 		if newMonSucceeded {
 			// do nothing if the new mon was started successfully, the deployment will anyway be deleted
 			return
 		}
+		logger.Warningf("failover of mon %q unsuccessful, cleaning up failover mon %q", name, m.DaemonName)
 		if err := c.updateMonDeploymentReplica(name, true); err != nil {
 			// attempt to continue even if the bad mon could not be restarted
 			logger.Warningf("failed to restart failed mon %q after new mon wouldn't start. %v", name, err)
 		}
-	}()
-
-	// remove the failed mon from a local list of the existing mons for finding a stretch zone
-	existingMons := c.clusterInfoToMonConfig(name)
-
-	zone, err := c.findAvailableZoneIfStretched(existingMons)
-	if err != nil {
-		return errors.Wrap(err, "failed to find available stretch zone")
-	}
-
-	// Start a new monitor
-	m := c.newMonConfig(c.maxMonID+1, zone)
-	logger.Infof("starting new mon: %+v", m)
+		if err := c.removeMonWithOptionalQuorum(m.DaemonName, newMonMightBeInQuorum); err != nil {
+			logger.Infof("failed to remove mon %q from quorum. %v", m.DaemonName, err)
+		}
 
-	mConf := []*monConfig{m}
+		// Make sure the maxMonID is reverted to its previous value
+		// The maxMonId is committed to a configmap immediately after the mon deployment
+		// is started, even though c.maxMonID is not incremented until the mon failover is successful
+		logger.Infof("reverting maxMonId to %d", c.maxMonID)
+		if err := c.commitMaxMonIDRequireIncrementing(c.maxMonID, false); err != nil {
```

Comment:
What happens if the operator crashes before this can do its thing? Is there a possibility that maxMonId might grow out of check if something in this function causes the crash?

Reply:
If the operator crashed:
- If the mon wasn't cleaned up, at next start the operator would try to get all the mons running (even the new one that wasn't properly aborted)
- As long as there is mon quorum, the mon health check will be started even if a mon or two are down
- The health check will remove any extra mon if there is an extra one in quorum
- After some time (~10 minutes), the health check will attempt to failover again if quorum is not fully healthy. The next mon id will be used.

If the failover crashes yet again, the operator would increase the maxMonId. But this is a very slow loop (at least 10-15 minutes), and there is really no limit on mon IDs, so it won't cause any permanent issue, we would just need to fix the hypothetical crashing bug.